### ISSUE TITLE
Note for AFML

### ISSUE THREAD

**Issue by baobach** [baobach/CQF_June_Cohort#1]
## Chapter 3

1. Fixed time horizon e.g. daily return, should be avoided because fix time interval doesn't exhibit good statistical properties.
2. take profit and stoploss limits are a function of the risks involved in a bet.
3. `span0 = 100` is the default value for span. This value is used to compute alpha decay value in the `ewm()` method: $\alpha = 2 / (span + 1)$, for $(span \geq 1)$. [Link](https://stackoverflow.com/questions/73797820/why-alpha-2-span-1-in-pandas-dataframe-ewm)
4. Daily volatility estimate is a standard deviation of exponential weighted moving of daily returns. This value will be used to set take profit and stop loss limits.
5. Triple barrier method is path dependent. Sampling path is $[t_{i,0}, t_{i,0}+h]$, where $h$ is the vertical barrier, $t_{i,1}$ is the timestamp of the first barrier touch - the return associated with the observed feature is $r_{t_{i,0}, t_{i,1}}$.

**Comment by baobach** [comment]
## Chapter 4

1. Probability of drawing 1 time and **NOT** selecting $i$ in $I$ elements is $\frac{I-1}{I}$, therefore, probability of drawing $I$ times and **NOT** selecting $i$ in $I$ elements is $(\frac{I-1}{I})^{I} = (1 - \frac{1}{I})^{I}$. Note: $\lim_{I \rightarrow \infty} (1 - I^{-1})^I = e^{-1}$. Following the same argument, max number of non-overlapping outcomes is $K \leq I$, then probability of not selecting an item $i$ after $I$ draws with replacement on a set of $I$ items is $(1-K^{-1})^I$, therefore $\lim_{I \rightarrow \infty} (1-I^{-1}) ^{I \frac{K}{I}} \approx e^{-\frac{K}{I}}$
7. $y_i$ is a label for an observed feature $X_i$ where $y_i$ was a function of price bars that occurred over an interval. 
8. `t1`: A pandas series with the timestamps of the vertical barriers
9. $1_{t,i} = 1$ if the is an overlap between $[t_{i,0}, t_{i,1}]$ and $[t-1,t]$ and equal $0$ otherwise.
10. Concurrent labels $y_i$ and $y_j$ at $t$ when both are a function of at least one common return. Number of labels concurrent at $t$: $c_t = \sum_{i=1}^{I} 1_{t,i}$.
11. Harmonic mean/average is reciprocal of arithmetic mean
12. Uniqueness of label $i$ at time $t$ is $u_{t,i} = 1_{t,i} c_t^{-1}$
13. Average Uniqueness of label $i$ is the average $u_{i,t}$ over the label's lifespan: 
$\overline u_i = \left(\sum_{t=1}^{T} u_{t,i}\right)  \left(\sum_{t=1}^{T} 1_{t,i}\right)^{-1}$.
14. Computing $\overline u_i$ requires information that is not available until a future time, but this is not a problem because the score only used on the training set in combination with label information. These information are not used for forecasting either, hence there is no information leakage. This procedure allows us to assign a score $[0,1]$ for each observed feature.
15. Standard Bootstrap aka draw sample with repetition, when sampling in-bag observation will be redundant and similar to out-of-bag observations.
16. Sequential Bootstrap updates the probability after each draw -> increase the uniqueness of the sample -> Approx IID.
17. We need to weight observations by some function of both uniqueness and absolute return. [More about log returns](https://quant.stackexchange.com/questions/70997/how-to-calculate-the-log-return-of-portfolio)
18. Time decay as markets evolve, we should weight relevant new observations.
19. To form an indicator matrix, the row of the matrix correspond to the index of returns which were used to label the data set and collumn correspond to samples.  Example: Taking a sample with 3 labels $y_0, y_1, y_2$, label $y_0$ is a function of return $r_{0,2}$ meaning we are using row $0, 1, 2$ to label sample $y_0$. The first column correspond to return index $0, 1, 2$ will have value 0. This process loop through 2D matrix and have complexity $O(n^2)$. [Hudson&Thame Blog](https://hudsonthames.org/bagging-in-financial-machine-learning-sequential-bootstrapping-python/)

**Comment by baobach** [comment]
## Chapter 5

1. Timeseries has memory, transformation such as returns, log-returns, etc. trying to find stationarity in the timeseries. This will strip away the memory which can have strong predictive power. This trade off is a dilemma.
2. Fix-width Window Fracdiff (FFD) approach shows that there is no need to give up all of the memory in order to gain stationarity.
3. Compute a cum sum of the time series, this guarantees that some order of differentiation is needed.
4. Compute the FFD(d) series for various $d \in [0,1]$.
5. Determine the minimum d such that the p-value of the ADF statistic on FFD falls below 5%
6. Use the FFD series as the predictive feature.
7. https://mathpages.com/home/kmath616/kmath616.htm

**Comment by baobach** [comment]
## Data Preparation Summary

1. Transform incoming `tick_data` to the desired bar: `volume_bar`, `dollar_bar`, etc
2. Form a series of cum-log-prices and apply FFD to find minimum d required
3. Apply CUSUM filter to the FFD we obtained from step 2
4. Form new feature matrix using the events selected by the CUSUM filter
5. Using triple barrier method with target = 2 * daily volatility and vertical barrier of 5 days
6. Using the sequential bootstrap method and determine sample weights, return weight.
7. Fit bagging classifiers of decision trees.

**Comment by baobach** [comment]
## Chapter 6

1. In financial application, Bagging is favored over Boosting. Bagging addresses overfitting, while boosting addresses underfitting. Overfitting is often a greater concern than underfitting, as it is not difficult to overfit an ML algorithm to financial data, because of the low signal-to-noise ratio.
2. If you use sklearn’s BaggingClassifier class to compute the out-of-bag accuracy, you should be aware of this [bug](https://github.com/scikit-learn/scikitlearn/issues/8933). One workaround is to rename the labels in integer sequential order.
3. Bagging’s main advantage is that it reduces forecasts’ variance, hence helping address overfitting.
4. Bagging classifier’s accuracy exceeds the average accuracy of the individual classifiers.
5. However, unlike boosting, bagging cannot improve the accuracy of poor classifiers: If the individual learners are poor classifiers $(p \ll \frac{1}{k})$, majority voting will still perform poorly (although with lower variance). For further analysis on this topic, check out Condorcet’s Jury Theorem.
6. Given that bagging algorithms can be parallelized, we are transforming a large
sequential task into many smaller ones that are run simultaneously. The early stopping will increase the variance of the outputs from the individual base estimators; however, that increase can be more than offset by the variance reduction associated with the bagging algorithm. You can control that reduction by adding more independent base estimators. Used in this way, bagging will allow you to achieve fast and robust estimates on extremely large datasets.

**Comment by baobach** [comment]
## Chapter 7 

1. K-Fold fails in finance is because observations cannot be assumed to be drawn from an IID process. A second reason for CV’s failure is that the testing set is used multiple times in the process of developing a model, leading to multiple testing and selection bias.
2. Because of the serial correlation i.e. $X_t \approx X_{t+1}$ and labels are derived from overlapping datapoints i.e. $Y_t \approx Y_{t+1}$, by placing $t$ and $t+1$ in different sets, information is leaked.
3. Solution is to Purge the K-fold CV i.e. removes overlap timestamps that contains leakage. We will determine that there is informational overlap between two observations $i$ and $j$ whenever $Y_i$ and $Y_j$ are concurrent (see Chapter 4, Section 4.3).
4. When purging is not able to prevent all leakage, we can impose an Embargo on training observations after every test set. An embargo period $h \approx 0.01 T$ suffices to prevent all leakage.

**Comment by baobach** [comment]
## Chapter 8
1. Feature importance analysis should be performed BEFORE any backtest is carried out.
2. Once we have found what features are important, we can learn more by conducting a number of experiments: Are these features important all the time, or only in some specific environments? What triggers a change in importance over time? Can those regime switches be predicted? Are those important features also relevant to other related financial instruments? Are they relevant to other asset classes? What are the most relevant features across all financial instruments? What is the subset of features with the highest rank correlation across the entire investment universe?
3. Substitution effect takes place when the estimated importance of one feature is reduced by the presence of other related features.
4. Stacked method can be considered more reliable. It is not affected by randomness of splits which can cause disbalanced samples and instead estimates the dataset as a whole.
